src/main/js/utils/generate.js

```diff
@@ -5,58 +5,93 @@ import {
   if (courseList.length !== 0) {
     dispatch(startGenerating());
+    dispatch(selectSort(values.sortByGPA));
+
+    const shortCode = loadQuery || !values.genURL ? Promise.resolve(loadQuery.substring(2))
+      : fetchShortUrl(courseList, values);
+
+    const gpasLoading = values.sortByGPA ? getGPAMap(courseList) : Promise.reject();
+
     const filteredResults = filterCourses(values, courseList);
     dispatch(filteredCourses(filteredResults));
-    const filteredCourseList = filteredResults.map(result => (
-      result.filtered));
-    let schedules = [];
-    if (!filteredCourseList.find(list => list.length === 0)) {
-      const scheduleMaker = new ScheduleMaker(filteredCourseList, parseInt(values.gap, 10));
-      try {
-        schedules = scheduleMaker.makeSchedules();
-      } catch (e) {
-        schedules = e;
-      }
-    }
-    const queryObject = {
-      ...values,
-      c: courseList.map((course) => {
-        const name = `${course[0].subject}${course[0].courseNumber}`;
-        return JSON.stringify({
-          name,
-          ...(values.courses ? values.courses[`${name}${course.id}`] : {}),
-        });
-      }),
-    };
-    delete queryObject.courses;
-    (loadQuery ? Promise.resolve() : fetch(`${shortenUrl}?${stringify({ data: JSON.stringify(queryObject), prefix: 'generator' })}`))
-      .then(response => (loadQuery ? loadQuery.substring(2) : response.json()))
-      .then((result) => {
-        const querystring = `q=${result}`;
-        dispatch(startRedirect({
-          pathname: '/generator',
-          hash: schedules.length > 0 ? '#schedules' : '#timetable',
-          search: result ? querystring : '',
-        }));
-      })
-      .catch(() => {
-        dispatch(startRedirect({
-          pathname: '/generator',
-          hash: schedules.length > 0 ? '#schedules' : '#timetable',
-          search: '',
+    const filteredCourseList = filteredResults.map(result => result.filtered);
+
+    gpasLoading
+      .then((gradeMap) => {
+        dispatch(addGrades(gradeMap));
+        filteredCourseList.forEach(list => list.sort((course1, course2) => {
+          const name = `${course1.subject}${course1.courseNumber}`;
+          const instructor1 = getInstructorLastName(course1.instructor); // Backend only holds instructors' last name
+          const instructor2 = getInstructorLastName(course2.instructor);
+
+          if (!gradeMap[name]) return 0;
+
+          if (!gradeMap[name][instructor1] && !gradeMap[name][instructor2]) return 0;
+          if (!gradeMap[name][instructor1]) return 1;
+          if (!gradeMap[name][instructor2]) return -1;
+          return gradeMap[name][instructor2] - gradeMap[name][instructor1];
         }));
       })
-      .then(() => {
-        dispatch(endGenerating(schedules));
+      .catch(error => error)
+      .finally(() => {
+        const schedules = genSchedules(filteredCourseList, values.gap);
+
+        if (values.sortByGPA) {
```

Comment:
since schedules could be an error object

`if (schedules && values.sortByGPA)`

Reply:
Fixed.